pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.symbols;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasProperty;
+import static org.hamcrest.Matchers.hasSize;
+
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import net.sourceforge.pmd.PMDConfiguration;
+import net.sourceforge.pmd.internal.util.ClasspathClassLoader;
+import net.sourceforge.pmd.lang.java.types.JClassType;
+import net.sourceforge.pmd.lang.java.types.TypeSystem;
+
+/**
+ * @author Clément Fournier
+ */
+class ClassLoadingChildFirstTest {
+
+
+    /**
+     * In this test I packed a jar with a custom version of {@link Void}.
+     * The test asserts that when putting that jar on the auxclasspath,
+     * a {@link TypeSystem} prefers that custom class to the one on the
+     * bootstrap classpath. The functionality under test is actually in
+     * {@link ClasspathClassLoader}.
+     */
+    @Test
+    void testClassLoading() {
+        String jarPath = getClass().getPackage().getName().replace('.', '/') + "/custom_java_lang.jar";
+        URL url = Thread.currentThread().getContextClassLoader().getResource(jarPath);
+        Path file = Paths.get(url.getPath());
+
+        PMDConfiguration config = new PMDConfiguration();
+        config.prependAuxClasspath(file.toAbsolutePath().toString());
+
+        TypeSystem typeSystem = TypeSystem.usingClassLoaderClasspath(config.getClassLoader());
+
+        JClassType voidClass = typeSystem.BOXED_VOID;
+        List<JMethodSymbol> declaredMethods = voidClass.getSymbol().getDeclaredMethods();
+        assertThat(declaredMethods, hasSize(1));
+        assertThat(declaredMethods.get(0), hasProperty("simpleName", equalTo("customMethodOnJavaLangVoid")));
```

Comment:
Maybe we should at least provide the sources of custom_java_lang.jar here as a comment. If I understand it correctly, it looks like this:

```java
package java.lang;
public final class Void {
    public static final Class<Void> TYPE = (Class<Void>) Class.getPrimitiveClass("void");
    private Void() {}
    public void customMethodOnJavaLangVoid() {}
}
```

Maybe we could simply add the source additionally into the jar?

Reply:
I see, you already added the sources to the jar. Perfect!